### ISSUE TITLE
SpringViewDisplayPostProcessor incompatible with bean reconfiguration and JRebel, etc.

### ISSUE THREAD

**Issue by davidtemplin** [vaadin/spring#200]
The `SpringViewDisplayPostProcessor` bean registers a bean of type `SpringViewDisplayRegistrationBean` each time the post-processor executes. Normally, this is acceptable, but it certain circumstances, e.g. when using JRebel for development, an `@SpringView`-annotated bean can be reconfigured, causing the post-processor to execute again and register a duplicate bean definition for the `SpringViewDisplayRegistrationBean` with the same `SpringViewDisplayRegistrationBean.beanClass` value. This will cause subsequent bean injection to fail with a `NoUniqueBeanDefinitionException` since there is now a duplicate bean definition.

In my experience it is very common for Vaadin developers to utilize JRebel to accelerate development. This problem renders JRebel useless without the workaround described below.

To reproduce: create any Vaadin Spring project with at least one `@SpringView`-annotated view and at least one `@SpringViewDisplay`-annotated component and use the JRebel agent to reload the view in a running application.

The following workaround alleviates the issue:

1. Do not use `@EnableVaadinNavigation`
2. Instead, explicitly register the following beans in your Spring configuration, e.g.:
```java
    @Bean
    @UIScope
    public SpringNavigator vaadinNavigator() {
        return new SpringNavigator();
    }

    @Bean
    public static SpringViewDisplayPostProcessor springViewDisplayPostProcessor() {
        return new WorkaroundSpringViewDisplayPostProcessor();
    }
```
3. Create the following two classes:

**WorkaroundSpringViewDisplayPostProcessor:**
```java
import com.vaadin.spring.internal.SpringViewDisplayPostProcessor;
import com.vaadin.spring.internal.UIScopeImpl;
import org.springframework.beans.factory.BeanDefinitionStoreException;
import org.springframework.beans.factory.config.BeanDefinition;
import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
import org.springframework.beans.factory.support.AbstractBeanDefinition;
import org.springframework.beans.factory.support.BeanDefinitionBuilder;
import org.springframework.beans.factory.support.BeanDefinitionRegistry;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ConfigurableApplicationContext;

public class WorkaroundSpringViewDisplayPostProcessor extends SpringViewDisplayPostProcessor {

    private ApplicationContext applicationContext;

    @Override
    public void setApplicationContext(ApplicationContext applicationContext) {
        super.setApplicationContext(applicationContext);
        this.applicationContext = applicationContext;
    }

    @Override
    protected void registerSpringViewDisplayBean(Class<?> clazz) {
        BeanDefinitionRegistry registry = null;
        if (applicationContext instanceof BeanDefinitionRegistry) {
            registry = (BeanDefinitionRegistry) applicationContext;
        } else if (applicationContext instanceof ConfigurableApplicationContext) {
            ConfigurableListableBeanFactory beanFactory = ((ConfigurableApplicationContext) applicationContext).getBeanFactory();
            if (beanFactory instanceof BeanDefinitionRegistry)
                registry = (BeanDefinitionRegistry) beanFactory;
        }
        if (registry == null) {
            throw new BeanDefinitionStoreException("BeanDefinitionRegistry is not accessible");
        }

        BeanDefinitionBuilder builder = BeanDefinitionBuilder
          .genericBeanDefinition(WorkaroundSpringViewDisplayRegistrationBean.class);

        // information needed to extract the values from the current UI scoped
        // beans
        builder.addPropertyValue("beanClass", clazz);

        builder.setScope(UIScopeImpl.VAADIN_UI_SCOPE_NAME);
        builder.setRole(BeanDefinition.ROLE_SUPPORT);
        AbstractBeanDefinition beanDefinition = builder.getBeanDefinition();
        String name = getBeanNameGenerator().generateBeanName(beanDefinition,
          registry);
        final boolean alreadyRegistered = applicationContext
          .getBeansOfType(WorkaroundSpringViewDisplayRegistrationBean.class)
          .values()
          .stream()
          .map(WorkaroundSpringViewDisplayRegistrationBean::getBeanClass)
          .anyMatch(clazz::equals);
        if (!alreadyRegistered)
            registry.registerBeanDefinition(name, beanDefinition);
    }

    @Override
    protected void registerSpringViewDisplayBean(String beanName) {
        BeanDefinitionRegistry registry = (BeanDefinitionRegistry) applicationContext;
        BeanDefinitionBuilder builder = BeanDefinitionBuilder
          .genericBeanDefinition(WorkaroundSpringViewDisplayRegistrationBean.class);

        // information needed to extract the values from the current UI scoped
        // beans
        builder.addPropertyValue("beanName", beanName);

        builder.setScope(UIScopeImpl.VAADIN_UI_SCOPE_NAME);
        builder.setRole(BeanDefinition.ROLE_SUPPORT);
        AbstractBeanDefinition beanDefinition = builder.getBeanDefinition();
        String name = getBeanNameGenerator().generateBeanName(beanDefinition,
          registry);
        final boolean alreadyRegistered = applicationContext
          .getBeansOfType(WorkaroundSpringViewDisplayRegistrationBean.class)
          .values()
          .stream()
          .map(WorkaroundSpringViewDisplayRegistrationBean::getBeanName)
          .anyMatch(beanName::equals);
        if (!alreadyRegistered)
            registry.registerBeanDefinition(name, beanDefinition);
    }
}
```
**WorkaroundSpringViewDisplayRegistrationBean:**
```java
public class WorkaroundSpringViewDisplayRegistrationBean extends SpringViewDisplayRegistrationBean {

    private Class<?> beanClass;
    private String beanName;

    @Override
    public void setBeanClass(Class<?> beanClass) {
        super.setBeanClass(beanClass);
        this.beanClass = beanClass;
    }

    @Override
    public void setBeanName(String beanName) {
        super.setBeanName(beanName);
        this.beanName = beanName;
    }

    public Class<?> getBeanClass() {
        return beanClass;
    }

    public String getBeanName() {
        return beanName;
    }
}
```
**Suggested patch:**

I'm not sure if it is appropriate, but the following patch sufficed as a temporary workaround:

**In SpringViewDisplayPostProcessor:**
```diff
--- a/vaadin-spring/src/main/java/com/vaadin/spring/internal/SpringViewDisplayPostProcessor.java
+++ b/vaadin-spring/src/main/java/com/vaadin/spring/internal/SpringViewDisplayPostProcessor.java
@@ -116,7 +116,14 @@ public class SpringViewDisplayPostProcessor implements BeanPostProcessor,
         AbstractBeanDefinition beanDefinition = builder.getBeanDefinition();
         String name = getBeanNameGenerator().generateBeanName(beanDefinition,
                 registry);
-        registry.registerBeanDefinition(name, beanDefinition);
+        final boolean alreadyRegistered = applicationContext
+          .getBeansOfType(WorkaroundSpringViewDisplayRegistrationBean.class)
+          .values()
+          .stream()
+          .map(WorkaroundSpringViewDisplayRegistrationBean::getBeanClass)
+          .anyMatch(clazz::equals);
+        if (!alreadyRegistered)
+            registry.registerBeanDefinition(name, beanDefinition);
     }
 
     /**
@@ -140,7 +147,14 @@ public class SpringViewDisplayPostProcessor implements BeanPostProcessor,
         AbstractBeanDefinition beanDefinition = builder.getBeanDefinition();
         String name = getBeanNameGenerator().generateBeanName(beanDefinition,
                 registry);
-        registry.registerBeanDefinition(name, beanDefinition);
+        final boolean alreadyRegistered = applicationContext
+          .getBeansOfType(WorkaroundSpringViewDisplayRegistrationBean.class)
+          .values()
+          .stream()
+          .map(WorkaroundSpringViewDisplayRegistrationBean::getBeanName)
+          .anyMatch(beanName::equals);
+        if (!alreadyRegistered)
+            registry.registerBeanDefinition(name, beanDefinition);
     }
```

**In SpringViewDisplayRegistrationBean:**

```diff
--- a/vaadin-spring/src/main/java/com/vaadin/spring/internal/SpringViewDisplayRegistrationBean.java
+++ b/vaadin-spring/src/main/java/com/vaadin/spring/internal/SpringViewDisplayRegistrationBean.java
@@ -45,6 +45,16 @@ public class SpringViewDisplayRegistrationBean implements Serializable {
     }
 
     /**
+     * Get the class of the bean that has the view display annotation.
+     *
+     * @return class of the bean that contains the SpringViewDisplay
+     *         annotation or has it directly on the class
+     */
+    public Class<?> getBeanClass() {
+        return this.beanClass;
+    }
+
+    /**
      * Set the class of the bean that has the view display annotation. Either
      * this method or {@link #setBeanName(String)} should be called.
      *
@@ -57,6 +67,15 @@ public class SpringViewDisplayRegistrationBean implements Serializable {
     }
 
     /**
+     * Get the name of the bean that has the view display annotation.
+     *
+     * @return name of the bean that has the SpringViewDisplay annotation
+     */
+    public String getBeanName() {
+        return this.beanName;
+    }
+
+    /**
      * Set the name of the bean that has the view display annotation. Either
```

**Comment by markathomas** [comment]
+1 - please fix

**Comment by hesara** [comment]
Fixed for 1.2.1 and 2.0.2
